src/evict.c

```diff
@@ -586,33 +587,32 @@ int performEvictions(void) {
             server.maxmemory_policy == MAXMEMORY_VOLATILE_TTL)
         {
             struct evictionPoolEntry *pool = EvictionPoolLRU;
+            dbKeyType keyType = (server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS ?
+                                 DB_MAIN : DB_EXPIRES);
 
             while (bestkey == NULL) {
-                unsigned long total_keys = 0, keys;
+                unsigned long total_keys = 0;
 
                 /* We don't want to make local-db choices when expiring keys,
                  * so to start populate the eviction pool sampling keys from
                  * every DB. */
                 for (i = 0; i < server.dbnum; i++) {
                     db = server.db+i;
+                    unsigned long sampled_keys = 0;
+                    unsigned long current_db_keys = dbSize(db, keyType);
+                    if (current_db_keys == 0) continue;
+
+                    total_keys += current_db_keys;
                     do {
-                        int slot = 0;
-                        if (server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS) {
-                            slot = getFairRandomSlot(db, DB_MAIN);
-                            dict = db->dict[slot];
-                        } else {
-                            slot = getFairRandomSlot(db, DB_EXPIRES);
-                            dict = db->expires[slot];
-                        }
-                        if ((keys = dictSize(dict)) != 0) {
-                            evictionPoolPopulate(i, slot, dict, db, pool);
-                            total_keys += keys;
-                        }
-                    /* Since keys are distributed across smaller slot-specific dictionaries in cluster mode, we may need to
-                     * visit more than one dictionary in order to populate required number of samples into eviction pool. */
-                    } while (server.cluster_enabled && keys != 0 && server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS &&
-                        total_keys < (unsigned long) server.maxmemory_samples
-                    );
+                        int slot = getFairRandomSlot(db, keyType);
+                        dict = (keyType == DB_MAIN ? db->dict[slot] : db->expires[slot]);
+                        sampled_keys += evictionPoolPopulate(i, slot, dict, db->dict[slot], pool);
```

Comment:
IIUC for non CME we would be trying to pick keys from the same dict multiple times. Is there a risk of picking lot's of duplicate keys for eviction? 

Should we do a de-dup while doing the insertion sort in `evictionPoolPopulate` ?

Reply:
before #11695 we didn't have any loop here, maybe in standalone we should drop this loop altogether?

Reply:
I hope to decouple eviction from the running mode as much as possible. I have added statistics for non-empty slots and used it as another termination condition. This means that the number of iterations should not exceed the number of non-empty slots. Additionally, I think this variable may be useful for other systems as well.

Reply:
the change you did seems good to me.
i wanna make sure that @hpatro think it's good enough (we could still sample keys from the same dict multiple times).
also, please have a look at https://github.com/redis/redis/pull/12781#discussion_r1407535746 and ack.